pint_xarray/checking.py

```diff
@@ -0,0 +1,237 @@
+import functools
+import inspect
+from inspect import Parameter
+
+from pint import Quantity
+from xarray import DataArray, Dataset
+
+from .accessors import PintDataArrayAccessor  # noqa
+
+
+def expects(*args_units, return_units=None, **kwargs_units):
+    """
+    Decorator which ensures the inputs and outputs of the decorated
+    function are expressed in the expected units.
+
+    Arguments to the decorated function are checked for the specified
+    units, converting to those units if necessary, and then stripped
+    of their units before being passed into the undecorated
+    function. Therefore the undecorated function should expect
+    unquantified DataArrays, Datasets, or numpy-like arrays, but with
+    the values expressed in specific units.
+
+    .. note::
+        The coordinates of input DataArrays are not checked, only the
+        data. So if your decorated function uses coordinates and you
+        wish to check their units, you should pass the coordinates of
+        interest as separate arguments.
+
+    Parameters
+    ----------
+    func : callable
+        Function to decorate, which accepts zero or more
+        xarray.DataArrays or numpy-like arrays as inputs, and may
+        optionally return one or more xarray.DataArrays or numpy-like
+        arrays.
+    *args_units : unit-like or mapping of hashable to unit-like, optional
+        Units to expect for each positional argument given to func.
+
+        The decorator will first check that arguments passed to the
+        decorated function possess these specific units (or will
+        attempt to convert the argument to these units), then will
+        strip the units before passing the magnitude to the wrapped
+        function.
+
+        A value of None indicates not to check that argument for units
+        (suitable for flags and other non-data arguments).
```

Comment:
that's not consistent with the rest of the library, where `None` means "not a quantity" (i.e. just the magnitude). I think that would fit here, too.

Reply:
> `None` means "not a quantity" (i.e. just the magnitude)

Sorry, I'm not following - what exactly are you suggesting I do here? If I change `None` to mean "take the `.magnitude` of that argument` then what should I when the argument is a string or something without a `.magnitude`?

Reply:
what I meant with this is that `.quantify` will convert the data to:
* `str` / `Unit` → create a `Quantity` object
* `None` → return the data unchanged

I was probably a bit too tired yesterday, but reading it now it sounds like "use `None` for values that don't represent a `Quantity`, which is fine for now.